File: vi/codes/swift/chapter_stack_and_queue/linkedlist_stack.swift
/**
 * File: linkedlist_stack.swift
 * Created Time: 2023-01-09
 * Author: nuomi1 (nuomi1@qq.com)
 */

import utils

/* Ngăn xếp dựa trên cách triển khai danh sách liên kết */
class LinkedListStack {
    private var _peek: ListNode? // Dùng nút đầu làm đỉnh ngăn xếp
    private var _size: Int // Độ dài ngăn xếp

    init() {
        _size = 0
    }

    /* Lấy độ dài của ngăn xếp */
    func size() -> Int {
        _size
    }

    /* Kiểm tra ngăn xếp có rỗng không */
    func isEmpty() -> Bool {
        size() == 0
    }

    /* Đẩy vào ngăn xếp */
    func push(num: Int) {
        let node = ListNode(x: num)
        node.next = _peek
        _peek = node
        _size += 1
    }

    /* Lấy ra khỏi ngăn xếp */
    @discardableResult
    func pop() -> Int {
        let num = peek()
        _peek = _peek?.next
        _size -= 1
        return num
    }

    /* Trả về danh sách để in ra */
    func peek() -> Int {
        if isEmpty() {
            fatalError("Stack is empty")
        }
        return _peek!.val
    }

    /* Chuyển List thành Array và trả về */
    func toArray() -> [Int] {
        var node = _peek
        var res = Array(repeating: 0, count: size())
        for i in res.indices.reversed() {
            res[i] = node!.val
            node = node?.next
        }
        return res
    }
}

@main
enum _LinkedListStack {
    /* Mã điều khiển */
    static func main() {
        /* Truy cập phần tử ở đỉnh ngăn xếp */
        let stack = LinkedListStack()

        /* Đẩy các phần tử vào ngăn xếp */
        stack.push(num: 1)
        stack.push(num: 3)
        stack.push(num: 2)
        stack.push(num: 5)
        stack.push(num: 4)
        print("Stack stack = \(stack.toArray())")

        /* Trả về danh sách để in ra */
        let peek = stack.peek()
        print("Top element peek = \(peek)")

        /* Lấy phần tử ra khỏi ngăn xếp */
        let pop = stack.pop()
        print("Pop element pop = \(pop), after pop stack = \(stack.toArray())")

        /* Lấy độ dài của ngăn xếp */
        let size = stack.size()
        print("Stack length size = \(size)")

        /* Kiểm tra rỗng */
        let isEmpty = stack.isEmpty()
        print("Is stack empty = \(isEmpty)")
    }
}

File: vi/codes/c/chapter_stack_and_queue/linkedlist_stack.c
/**
 * File: linkedlist_stack.c
 * Created Time: 2023-01-12
 * Author: Zero (glj0@outlook.com)
 */

#include "../utils/common.h"

/* Ngăn xếp cài đặt dựa trên danh sách liên kết */
typedef struct {
    ListNode *top; // Dùng nút đầu làm đỉnh ngăn xếp
    int size;      // Độ dài ngăn xếp
} LinkedListStack;

/* Hàm khởi tạo */
LinkedListStack *newLinkedListStack() {
    LinkedListStack *s = malloc(sizeof(LinkedListStack));
    s->top = NULL;
    s->size = 0;
    return s;
}

/* Hàm hủy */
void delLinkedListStack(LinkedListStack *s) {
    while (s->top) {
        ListNode *n = s->top->next;
        free(s->top);
        s->top = n;
    }
    free(s);
}

/* Lấy độ dài của ngăn xếp */
int size(LinkedListStack *s) {
    return s->size;
}

/* Kiểm tra ngăn xếp có rỗng không */
bool isEmpty(LinkedListStack *s) {
    return size(s) == 0;
}

/* Đẩy vào ngăn xếp */
void push(LinkedListStack *s, int num) {
    ListNode *node = (ListNode *)malloc(sizeof(ListNode));
    node->next = s->top; // Cập nhật trường con trỏ của nút mới
    node->val = num;     // Cập nhật trường dữ liệu của nút mới
    s->top = node;       // Cập nhật đỉnh ngăn xếp
    s->size++;           // Cập nhật kích thước ngăn xếp
}

/* Truy cập phần tử đỉnh ngăn xếp */
int peek(LinkedListStack *s) {
    if (s->size == 0) {
        printf("Stack is empty\n");
        return INT_MAX;
    }
    return s->top->val;
}

/* Lấy ra khỏi ngăn xếp */
int pop(LinkedListStack *s) {
    int val = peek(s);
    ListNode *tmp = s->top;
    s->top = s->top->next;
    // Giải phóng bộ nhớ
    free(tmp);
    s->size--;
    return val;
}

/* Đoạn mã điều khiển chính */
int main() {
    /* Khởi tạo ngăn xếp */
    LinkedListStack *stack = newLinkedListStack();

    /* Đẩy phần tử vào ngăn xếp */
    push(stack, 1);
    push(stack, 3);
    push(stack, 2);
    push(stack, 5);
    push(stack, 4);

    printf("Stack stack = ");
    printLinkedList(stack->top);

    /* Truy cập phần tử đỉnh ngăn xếp */
    int val = peek(stack);
    printf("Top element top = %d\r\n", val);

    /* Lấy phần tử ra khỏi ngăn xếp */
    val = pop(stack);
    printf("Pop element pop = %d, stack after pop = ", val);
    printLinkedList(stack->top);

    /* Lấy độ dài của ngăn xếp */
    printf("Stack size = %d\n", size(stack));

    /* Kiểm tra có rỗng không */
    bool empty = isEmpty(stack);
    printf("Is stack empty = %s\n", empty ? "true" : "false");

    // Giải phóng bộ nhớ
    delLinkedListStack(stack);

    return 0;
}
